docs/src/proposals/traced-transaction-construction.md

```diff
@@ -0,0 +1,247 @@
+   }
+
+   transfer_counter.count += 1;
+   ...
+}
+```
+
+### example: sharding
+
+
+```rust
+pub fn process_place_order(
+  program_id: &Pubkey,     
+  market_id: &Pubkey,
+  order: &Pubkey,
+) {
+  let sharded_event_queue_address = let counter_address = Pubkey::find_program_address(
+      &[  
+          &market_id.to_bytes(),
+          &[ourder_to_bytes()[0] & 0xf0],
+      ],
+      program_id,
+  );
+  let event_queue = access_as_unsigned(sharded_event_queue_address, READ_WRITE);  
+
+  // crank tx will merge all event queues later
+}
+```
+
+### example: instruction versioning across redeploy
+
+```rust
+pub fn check_instruction_version_and_load_data() {
+   // this is newly added after the redeploy
+   if access_as_unsigned(declare!("OldVersion111111111111111111111111111111111111"), READ).is_ok() {
+     // it seems that this transaction is construction by simulation with pre-redeploy program executable.
+     return Err("we changed internal data loading implementation. to avoid unexpected malfunction retry");
+   }
+
+   // this is changed from OldVersion1111
+   access_as_unsigned(declare!("NewVersion111111111111111111111111111111111111"), READ)?
+
+   access_as_unsigned(...);
+}
+```
+
+
+## RPC changes
+
+`pre_balances` and `post_balances` needs to be updated?
+
+### sample return from `simulateTransaction` (or define `traceTransaction` for the compatibility??):
+
+```yml
+# these somewhat verbose formatting is for future expansion
+transaction:
+  instructions
+    accesses:
+      Bm9UhRBysjeT35ukzriheHKr76yiM1fPyEBMMw2yT9v1: {read: true, write: false}      
+      J8H1Gsvnm2CY5oVa1zw1LHHbfvtone8ZGsVBzjkkPbpy: {read: true, write: true}      
+    instructions:
+      accesses: []
+      instruction_type: spl_token_transfer
+      instructions: [...]
+  accesses: [...]
+  pre_balances/pre_token_balances
+  post_balances/post_token_balances
+  token_token_approvals: ???
+```
+
+
+
+also, page-indexed account can be incorporated to find compact combination of existing account indexes to match the required account list
+
+## downside
+
+race condition around program deploy (= possible account loading change) => just retry also seek to the ix versioning for perfection
+
+(traced program just load dummy addresses for tx versionsing for completeness)
+
+## careful consideration
+
+prevent program  detect to be simulated or really-executed
+
+
+## Other Proposals / alternative implementation
+
+### customized entrypoint?
+
+breaks compatibility?
+
+### Run these before replaying stage on validator
+
+unavoidable latency.  (and scheduling difficulty).
+The proposed implementation doesn't have such a downside
+
+## Future possible expansions
+
+- Wallet account balance delta check: append VerifyTokenBalances ix at last and pass the expected balances of spl-tokens or hash of it (also with delegation status).
+- Browser-side bpf program simualtion for server-less architecture with online wasm translation
```

Comment:
this is only possible because solana tx execution doesn't touch the whole account state in the chain. :)